feat: Add Prettier

willdurand opened the pull request:
Fixes https://github.com/mozilla/sign-addon/issues/328
Fixes #331 

---

This will conflict with some existing patches but I made 2 distinct commits to ease the rebase step in these other patches. In short, one can cherry-pick the first commit, apply it to their branch (before their commits), then run prettier. After that, remove the commit and rebase on top of master (once this patch is merged).

kumar303 commented on .prettierrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "arrowParens": "always",
+  "singleQuote": true,
```
I think it would be better to stick to the prettier defaults whenever possible. In other words, if we don't *need* to change a default to match some legacy code then let's not. In this case, the code already uses mostly double quotes so we should stick to that.

Looking back, I think it was a mistake to begin code-manager with non-default prettier options. In `addons-frontend`, it made sense because it resulted in a lighter diff, which makes `git blame` more effective.

kumar303 commented on .prettierrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "arrowParens": "always",
+  "singleQuote": true,
+  "trailingComma": "all",
```
I don't see any good reason for why we should change this default. In fact, `all` is annoying because it sets you up for a syntax error in some cases when you move code from a function call into a function body. I'm not sure why we ever used it on `addons-frontend` in the first place -- maybe to lighten the diff?

kumar303 commented:
> @kumar303 we use this config in _all_ our js projects now

Is that really the goal? I thought the original goal of overriding defaults was to match legacy code? 

I don't think it makes sense to *change* a prettier config for any project currently using prettier because it harms git blamability (`code-manager` is what it is, for example). I do think it's helpful to re-evaluate our application of prettier in new projects because it's our one shot at getting it right. 

Overriding prettier defaults erodes compatibility with prettier tools. Not all of them have strong support for a custom prettier config. I mean, they *should* but a non-default prettier still introduces complications.

willdurand commented:
> Is that really the goal? I thought the original goal of overriding defaults was to match legacy code?

mm, I thought it was to match our conventions.

> I don't think it makes sense to _change_ a prettier config for any project currently using prettier because it harms git blamability (`code-manager` is what it is, for example). I do think it's helpful to re-evaluate our application of prettier in new projects because it's our one shot at getting it right.
>
> Overriding prettier defaults erodes compatibility with prettier tools. Not all of them have strong support for a custom prettier config. I mean, they _should_ but a non-default prettier still introduces complications.

Which tool exactly?

I would really like to keep all our codebases consistent (and there is a lot more now than ever). I don't want to re-train my eyes every time I touch one of our js projects, which I have to do almost every day.

kumar303 commented:
> Which tool exactly?

It's just easier to run `prettier` than to run it with a config because it's non-trivial to locate the nearest config. I mean, I know that vscode can do it fine and there are plenty of vim tricks to get it working.

> I would really like to keep all our codebases consistent (and there is a lot more now than ever). I don't want to re-train my eyes every time I touch one of our js projects, which I have to do almost every day.

That's a fair point.

I am fine with using one config for all projects. It's actually easier for me because I use a dumb vim command for formatting (it doesn't know how to discover configs).

I guess we're stuck with a non-default prettier, heh. I didn't really think that was the original goal but striving for a common config across all projects seems fine and it would be *more* disruptive to change the existing configs.

willdurand commented:
> It's just easier to run `prettier` than to run it with a config because it's non-trivial to locate the nearest config. I mean, I know that vscode can do it fine and there are plenty of vim tricks to get it working.

I use this for `vim`: https://github.com/willdurand/dotfiles/blob/0ae452464d6c9a5b48551a60e56ee6377cb9ee63/vim/ftplugin/javascript.vim#L21-L22. It picks the config in the project's root directory because I start `vim` from that directory.

willdurand commented:
> I guess we're stuck with a non-default prettier, heh.

It might still be something to consider in the future. Sure it is not ideal to re-apply prettier on all our codebases, but that would be needed once.

As for `git blame`, I use this: https://jacob.hoffman-andrews.com/README/2018/12/07/chasing-git-blame.html

kumar303 commented on .eslintrc:
```diff
@@ -19,6 +19,10 @@
     "module": false,
     "require": false
   },
+  "extends": [
```
Do we need `amo` or [`amo/typescript`](https://github.com/mozilla/addons-code-manager/blob/master/.eslintrc#L4)? It could be done later but I was thinking maybe after prettier it wouldn't require code changes.

willdurand replied:
Earlier, I filed https://github.com/mozilla/sign-addon/issues/331.

kumar303 commented on .eslintrc:
```diff
@@ -54,7 +58,6 @@
     "no-with": 2,
-    "quotes": [2, "double", "avoid-escape"],
     "quote-props": [1, "consistent-as-needed"],
     "semi": [2, "always"],
```
A lot of these rules can probably be removed

willdurand replied:
Yes, I think so too. I filed https://github.com/mozilla/sign-addon/issues/331 for that. I only removed `quotes` because this rule was generating lots of errors.

kumar303 commented on package.json:
```diff
@@ -8,6 +8,9 @@
     "changelog-lint": "commitlint --from master",
     "lint": "grunt lint",
+    "prettier": "prettier --write '**'",
+    "prettier-ci": "prettier --list-different '**' || (echo '\n\nThis failure means you did not run `yarn prettier-dev` before committing\n\n' && exit 1)",
```
This can be a little annoying. A pre-commit script would be better. I've had a good experience with [husky](https://github.com/typicode/husky) but that could be a follow-up -- it was just a thought.

kumar303 commented on .prettierrc:
```diff
@@ -0,0 +1,6 @@
+{
```
I guess we really need a way to keep this in sync with a common config. Is there a way to use `extends`?

willdurand replied:
Maybe this? https://dev.to/fullstack_to/sharing-prettier-configuration-files-gba

willdurand replied:
I filed https://github.com/mozilla/addons/issues/1123

kumar303 commented on README.md:
```diff
@@ -153,50 +150,36 @@ and link back to `sign-addon`:
+- Create a [new release](https://github.com/mozilla/sign-addon/releases/new) and paste in the changelog Markdown. It may require some manual editing. For example, some commit messages might have been truncated. Title the github release after the new version you just added in the previous commit to `package.json` (example: `1.0.4`).
+- When you publish the release, github creates a tag. When TravisCI builds the tag, it will automatically publish the package to [npm](https://www.npmjs.com/package/sign-addon).
+
+<<<<<<< HEAD [typescript]: https://www.typescriptlang.org/ ======= [prettier]: https://prettier.io/
```
This looks like a merge conflict

willdurand replied:
ugh, I did not see it :/

willdurand replied:
Sorry about that

willdurand commented:
@kumar303 I simplified the eslint config and added the `amo` config/plugin. I removed most of the config. I think the `es6` env is not needed anymore and I am not sure about the `globals` either, but I did not remove these settings.

willdurand commented:
@kumar303 the last commit could totally be a new patch if you like. I tried not to change a lot of things but I had to fix most "no-param-reassign". Let me know what you think!

kumar303 commented on src/amo-client.js:
```diff
@@ -498,51 +712,62 @@ export class Client {
-  absoluteURL(url) {
-    if (!url.match(/^http/i)) {
-      url = this.apiUrlPrefix + url;
+  absoluteURL(_url) {
```
call it `urlString`, maybe?
